src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.String.cs

```diff
@@ -99,11 +92,8 @@ public static object Deserialize(string json, Type returnType, JsonSerializerOpt
                 var reader = new Utf8JsonReader(utf8, isFinalBlock: true, readerState);
                 result = ReadCore(returnType, options, ref reader);
 
-                if (reader.BytesConsumed != actualByteCount)
-                {
-                    ThrowHelper.ThrowJsonException_DeserializeDataRemaining(
-                        actualByteCount, actualByteCount - reader.BytesConsumed);
-                }
+                // The reader should have thrown if we have remaining bytes.
+                Debug.Assert(reader.BytesConsumed == actualByteCount);
```

Comment:
What happens if a user-defined converter doesn't read enough (or reads too much)?

Reply:
A custom converter should get the same results. We do have some checks in the (de)serializer for level and bytes consumed but that really just helps throw an exception earlier where the error occurred instead of in a later converter (built in or custom).

I was unable to get the original code from executing; it also shows up as not hit in code coverage.

The reader when isFinalBlock==true will throw when there is too much JSON.

The reader will also throw when too little:
```
'',' is invalid after a single JSON value. Expected end of data
```
```
Expected depth to be zero at the end of the JSON payload. There is an open JSON object or array that should be closed. 
```

Reply:
I was thinking something like this:
```C#
[Fact]
public static void NotEnoughData()
{
    string json = "{}   ";
    byte[] utf8 = Encoding.UTF8.GetBytes(json);
    var reader = new Utf8JsonReader(utf8); // isFinalBlock=true
    reader.Read(); // returns true
    reader.Read(); // returns true
    Assert.Equal(utf8.Length, reader.BytesConsumed); // Fails, utf8.Length = 5, reader.BytesConsumed = 2
    // Note, a third Read() which returns false is required
}
```

Reply:
I added an additional custom converter that leaves extra whitespace